feat: 풀이 글 리스트 보기(#31)

handseokjin opened the pull request:
- [x] 🔀 PR 제목의 형식을 잘 작성했나요? e.g. `[feat] PR을 등록한다.` 
- [x] 💯 테스트는 잘 통과했나요?
- [x] 🏗️ 빌드는 성공했나요?
- [x] 🧹 불필요한 코드는 제거했나요?
- [x] 💭 이슈는 등록했나요?
- [x] 🏷️ 라벨은 등록했나요?
- [x] 💻 git rebase를 사용했나요?

## 작업 내용
1. 풀이 글 목록 보기 기능 추가
   - 무한스크롤을 위한 페이징 처리 구현
   - 생성일자 순으로 내림차순 정렬 후 조회

2. 풀이 글 관련 요청 인증 여부 수정(SecurityConfig)
   - GET만 인증 없이 가능하도록 수정(목록 및 상세 조회)


## 스크린샷
※ DB에 저장된 값
 - 문제 1의 풀이 글 : 3개
 - 문제 2의 풀이 글 : 1개
<img width="775" alt="db에 저장된 값" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/1d7c9cfd-8d64-4f7e-a4d7-8ba1d8cacbd3"><br/><br/>

※ 스크린샷
1. 최초 요청(클라이언트에서 -100이라는 값으로 주기로 하였습니다)<br/>
  - 요청한 문제의 맨 처음 풀이 글부터 가져옵니다.
<img width="838" alt="최초 요청(-100)" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/991090d8-14bf-43e9-a37b-a0cea2fe28cb"><br/>

3. 최초 요청 시 받은 nextCursorId 값으로 다음 요청
<img width="857" alt="다음 요청 1" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/cea646e8-a0ff-4b7f-b5df-05b0ad9f7eec">
<img width="838" alt="다음 요청 2" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/363244ed-8fa2-4200-a67a-dee5563635bb">

4. 다음에 받은 nextCursorId 값으로 요청
  - 조회할 풀이 글이 없을 경우 nextCursorId 값 -1 반환
<img width="853" alt="요청 시 없을 경우 -1 반환" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/590711c5-bd88-49f0-8d4c-5d1bc2fdcafd">

5. nextCursorId 값 -1로 반환 시 조회 x
<img width="856" alt="-1으로 요청 시" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/a9ea1b4b-5bf8-4a62-b425-7d01ac0ccec0"><br/><br/>

※ 참고
  - 프론트에서의 편의를 위해 서버에서 받은 nextCursorId값을 그대로 전달해 주면 되도록 개발하였습니다.
  - 더 이상 조회할 데이터가 없음에도 조회하는 것은 비효율적이라 생각하여 nextCursorId 값이 -1일 경우에는 아예 조회 자체를 안 하도록 하였습니다.

## 주의사항

Closes #31

Commit:
feat: 풀이 글 리스트 보기(#31)

handseokjin commented:
jpa를 잘 모르다보니 쿼리 만든 거를 jpa로 하나하나 바꿔보면서 검증하느라 좀 오래 걸렸네요ㅠ(추가로 람다도..)

acisliver commented on src/main/java/com/nakaligoba/backend/global/config/SecurityConfig.java:
```diff
@@ -35,7 +36,8 @@ protected void configure(HttpSecurity http) throws Exception {
                 .authorizeRequests()
-                .antMatchers("/api/v1/auth/**", "/api/test", "/api/v1/problems/**").permitAll()
+                .antMatchers("/api/v1/auth/**", "/api/test").permitAll()
+                .antMatchers(HttpMethod.GET, "/api/v1/problems/**").permitAll()
```
GET 메서드만 허용하는 방법이 있었군요! 너무 좋습니다 👍

acisliver commented on src/main/java/com/nakaligoba/backend/solution/application/SolutionService.java:
```diff
@@ -57,4 +69,50 @@ public long createSolution(String writerEmail, long problemId, SolutionCreateReq
         return solution.getId();
     }
+
+    @Transactional
```
DB 조회만 하는 로직이라 readOnly = true를 붙여주는 것이 좋을 것 같습니다!

acisliver commented on src/main/java/com/nakaligoba/backend/solution/controller/SolutionController.java:
```diff
@@ -28,4 +29,11 @@ public ResponseEntity<Void> createSolution(@PathVariable("id") long id, @Request
     }
+
+    @GetMapping("/{id}/solutions")
+    public ResponseEntity<SolutionsResponse> readSolutions(@PathVariable("id") long id, @RequestParam long cursor, @RequestParam int size) {
```
파라미터 타입은 원시 자료형`long`이 아닌 래퍼 클래스`Long`를 사용하는 것이 좋습니다!
만약 만약 해당 값이 들어오지 않았다면 변수 초기화 과정에서 원시 자료형은 기본 값 `0L`을 넣을 것이고 래퍼 클래스는 `null`을 넣을 것이기 때문에 원시 자료형은 해당 값이 0으로 입력 된건지 0으로 초기화된 것인지 알 수 없습니다!

handseokjin replied:
아 그냥 당연하게 썼었는데 그렇게 되는군요, 그러면 long 부분 외에도 원시 자료형 부분을 전부 래퍼 클래스 바꿔두겠습니다!

acisliver reviewed (commented):
고생하셨습니다!

Commit:
chore: 리뷰 부분 수정

- DB 조회만 하는 부분에 readOnly=true 추가(SolutionService > readSolutions())
- 컨트롤러 내 인자가 원시 자료형인 부분 래퍼 클래스로 수정(SolutionController > readSolutions())

Commit:
chore: application-local.yaml > ddl-auto: create로 변경

gyujin-Kimm commented:
항상 깔끔하게 해결하시는 것을 보면서 정말 많은 것을 배워갑니다!👍
그리고 프론트까지 생각하시는 자세는 정말 본받겠습니다!👍
